cli/azd/pkg/azdo/pipeline.go

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azdo
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/environment"
+	"github.com/azure/azure-dev/cli/azd/pkg/infra/provisioning"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/microsoft/azure-devops-go-api/azuredevops"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/build"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/taskagent"
+)
+
+// Creates a variable to be associated with a Pipeline
+func createBuildDefinitionVariable(value string, isSecret bool, allowOverride bool) build.BuildDefinitionVariable {
+	return build.BuildDefinitionVariable{
+		AllowOverride: &allowOverride,
+		IsSecret:      &isSecret,
+		Value:         &value,
+	}
+}
+
+// returns the default agent queue. This is used to associate a Pipeline with a default agent pool queue
+func getAgentQueue(ctx context.Context, projectId string, connection *azuredevops.Connection) (*taskagent.TaskAgentQueue, error) {
+	client, err := taskagent.NewClient(ctx, connection)
+	if err != nil {
+		return nil, err
+	}
+	getAgentQueuesArgs := taskagent.GetAgentQueuesArgs{
+		Project: &projectId,
+	}
+	queues, err := client.GetAgentQueues(ctx, getAgentQueuesArgs)
+	if err != nil {
+		return nil, err
+	}
+	for _, queue := range *queues {
+		if *queue.Name == "Default" {
+			return &queue, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find a default agent queue in project %s", projectId)
+}
+
+// find pipeline by name
+func pipelineExists(
+	ctx context.Context,
+	client build.Client,
+	projectId *string,
+	pipelineName *string,
+) (bool, error) {
+	getDefinitionsArgs := build.GetDefinitionsArgs{
+		Project: projectId,
+		Name:    pipelineName,
+	}
+
+	buildDefinitionsResponse, err := client.GetDefinitions(ctx, getDefinitionsArgs)
+	if err != nil {
+		return false, err
+	}
+	buildDefinitions := buildDefinitionsResponse.Value
+	for _, definition := range buildDefinitions {
+		if *definition.Name == *pipelineName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// create a new Azure DevOps pipeline
+func CreatePipeline(
+	ctx context.Context,
+	projectId string,
+	name string,
+	repoName string,
+	connection *azuredevops.Connection,
+	credentials AzureServicePrincipalCredentials,
+	env *environment.Environment,
+	console input.Console,
+	provisioningProvider provisioning.Options) (*build.BuildDefinition, error) {
+
+	client, err := build.NewClient(ctx, connection)
+	if err != nil {
+		return nil, err
+	}
+
+	var exists bool = true
+	var count = 0
+	var maxTries = 4
+	for exists {
+		exists, err = pipelineExists(ctx, client, &projectId, &name)
+		if err != nil {
+			return nil, err
+		}
+		count = count + 1
+
+		if exists {
+			name = fmt.Sprintf("%s - %s (%d)", name, repoName, count)
+		} else {
+			continue
+		}
+
+		if count >= maxTries {
+			return nil, fmt.Errorf("error creating new pipeline")
+		}
+	}
```

Comment:
Consider using the `retry` package that is used from other areas in the app.

Reply:
https://github.com/Azure/azure-dev/issues/756